### DISCUSSION TITLE
Is it possible to resolve DITA variables inside attributes?

### DISCUSSION THREAD

**Discussion by chrispy-snps** [dita-ot/dita-ot#4349] [Q&A] [answered]
We define product and release DITA variables in our top-level map:

```
  <keydef keys="Product">
    <topicmeta>
      <keywords>
        <keyword>My Product</keyword>
      </keywords>
    </topicmeta>
  </keydef>
  <keydef keys="Release">
    <topicmeta>
      <keywords>
        <keyword>2023.12</keyword>
      </keywords>
    </topicmeta>
  </keydef>
```

These variables are used in many places throughout the map.

Now I want to extract this information from published HTML5 in a standard way. No problem, I'll just set the `<prodinfo>` metadata fields to these variables, then reference the resulting HTML5 metadata fields. That's what metadata is for, right?

This works for `<prodname>` (because it's an element and thus can contain a DITA variable reference) but not for the `<vrm>` `@version` attribute (because it can't):

```
<map>
  <title>My Map</title>
  <topicmeta>
    <prodinfo>
      <prodname><keyword keyref="Product"/></prodname>  <!-- this works! -->
      <vrmlist>
        <vrm version="???"/>  <!-- so close... -->
      </vrmlist>
    </prodinfo>
  </topicmeta>
```

So close!! This would have worked out perfectly.

Is there some way to make this work with the existing processing, or will I need to resort to nonstandard DITA processing via a DITA-OT plugin?

**Comment by raducoravu** [comment]
@chrispy-snps we use entity references for that: https://github.com/oxygenxml/userguide/blob/master/DITA/maps/keydefs.ditamap

**Comment by jelovirt** [comment, marked as answer]
Unfortunately not, the bottle neck here is DITA spec itself.

**Comment by chrispy-snps** [comment]
Thanks @raducoravu and @jelovirt! I appreciate both answers.

**Reply by chrispy-snps** [reply]
I was able to implement a solution by setting `@version` to an empty string and including a `<data>` instance of the desired variable:

```
  <topicmeta>
    <prodinfo>
      <vrmlist>
        <vrm version=""/>  <!-- populated by com.synopsys.html5 during 'keyref' -->
      </vrmlist>
    </prodinfo>
    <data keyref="Release"/>
  </topicmeta>
```

then including the following template in our `mapref` plugin:

```
  <!-- if there is an empty <vrm> @version attribute, look for a
       <data keyref="Release"> element to populate it from -->
  <xsl:template match="*[contains-token(@class, 'topic/vrm')]/@version[. = '']">
    <xsl:attribute name="version" select="ancestor::*[contains-token(@class, 'map/topicmeta')]/*[contains-token(@class, 'topic/data')][matches(@keyref, '(^|\.)Release')]"/>
  </xsl:template>
```

Because the `mapref` stylesheet is called again after `KeyrefModule` resolves the key references, the template above can populate the value.

**Comment by lief-erickson** [comment]
@chrispy-snps, I'm curious how consistency will be ensured, especially over time and across people, including the future you. Forgetting the `<data>` seems easy to do. Is the intent to add a Schematron check to ensure the `<data>` element is present whenever there's an empty `@version`?

**Comment by chrispy-snps** [comment]
@lief-erickson - good idea! I added a Schematron rule to check for this:

```
  <!-- ensure a <data> element exists to populate an empty <vrm version=""> value -->
  <sch:pattern id="map_topicmeta">
    <sch:rule context="topicmeta/prodinfo/vrmlist/vrm" role="error">
      <sch:report test="@version = '' and not(../../../data[matches(@keyref, '(^|\.)Release')])">No &lt;data keyref="Release"/>&gt; element found to provide the data for this @version attribute.</sch:report>
    </sch:rule>
  </sch:pattern>
```

The Schematron rule allows for an optional keyscope prefix on the `<data keyref="[scope.]Release">` key reference. (In the root map of our online help collection, we use a `top` keyscope to prevent top-level keys from overriding book-level keys [as discussed here](https://dita-users.groups.io/g/main/topic/keyscope_question/103240470?p=,,,20,0,0,0::recentpostdate/sticky,,,20,2,0,103240470,previd%3D1703162672714830954,nextid%3D1695926464422817435&previd=1703162672714830954&nextid=1695926464422817435).)

I didn't include a Schematron quick-fix to insert the `<data>` element because for us, (1) the keyscope can vary depending on the map and (2) in most cases the whole `<prodinfo>`+`<data>` block of metadata will be inserted together.

Next, hopefully I can get the following enhancement request implemented so that the `<vrm @version>` value is included as HTML5 metadata in the top-level `index.html` file:

#4351: Generate HTML5 metadata attributes in the `index.html` TOC file
